Implement configuration for handling empty requests

Miciah opened the pull request:
#### Bump openshift/api for empty-requests API

Bump to https://github.com/openshift/api/commit/acb572771a9c4450a5267605e097ad0c091cf9ae to get the empty-requests API.

* `go.mod`: Bump.
* `go.sum`:
* `manifests/00-custom-resource-definition.yaml`:
* `pkg/manifests/bindata.go`:
* `vendor/*`: Regenerate.


#### Implement configuration for handling empty requests

Implement APIs for configuring handling of empty requests.

* `pkg/operator/controller/ingress/deployment.go` (`RouterDontLogNull`, `RouterHTTPIgnoreProbes`): New constants for the related environment variables.
(`desiredRouterDeployment`): Set the `ROUTER_DONT_LOG_NULL` and `ROUTER_HTTP_IGNORE_PROBES` environment variables as appropriate.
* `pkg/operator/controller/ingress/deployment_test.go` (`TestDesiredRouterDeployment`): Verify that `spec.logging.access.logEmptyRequests` and `spec.httpEmptyRequestsPolicy` have the expected effects.




----

Related to https://github.com/openshift/enhancements/pull/460.

Depends on https://github.com/openshift/api/pull/729.

Depends on https://github.com/openshift/router/pull/174.

Miciah commented:
/reopen
/remove-lifecycle rotten

Miciah commented:
Rebased.

Miciah commented:
Rebased again.

frobware commented on pkg/operator/controller/ingress/deployment_test.go:
```diff
@@ -287,6 +287,7 @@ func TestDesiredRouterDeployment(t *testing.T) {
 	checkDeploymentHasEnvVar(t, deployment, "ROUTER_CAPTURE_HTTP_REQUEST_HEADERS", false, "")
 	checkDeploymentHasEnvVar(t, deployment, "ROUTER_CAPTURE_HTTP_RESPONSE_HEADERS", false, "")
 	checkDeploymentHasEnvVar(t, deployment, "ROUTER_CAPTURE_HTTP_COOKIE", false, "")
+	checkDeploymentHasEnvVar(t, deployment, "ROUTER_DONT_LOG_NULL", false, "")
```
Can we reference the constant (`RouterDontLogNull`) for `"ROUTER_DONT_LOG_NULL"`? And ditto for "ROUTER_HTTP_IGNORE_PROBES".

Miciah replied:
Done.

Miciah commented:
Rebased on https://github.com/openshift/api/commit/acb572771a9c4450a5267605e097ad0c091cf9ae.

Miciah commented:
Rebased to fix a conflict from https://github.com/openshift/cluster-ingress-operator/pull/619.

frobware commented:
/lgtm

Miciah commented:
Rebased to resolve conflicts from https://github.com/openshift/cluster-ingress-operator/pull/452.

frobware commented:
/lgtm

Miciah commented:
One of the control-plane nodes never became ready.

/test e2e-aws-operator

Miciah commented:
`TestUniqueIdHeader` failed; looks like one of its test pods got deleted (garbage collected?) after the test was done with it but before the test got around to cleaning it up:

    === RUN   TestUniqueIdHeader
        operator_test.go:1914: failed to read output from pod unique-id-header-test-1: container "curl" in pod "unique-id-header-test-1" is waiting to start: ContainerCreating
        operator_test.go:1914: failed to read output from pod unique-id-header-test-1: container "curl" in pod "unique-id-header-test-1" is waiting to start: ContainerCreating
        operator_test.go:1926: found x-unique-id header from pod unique-id-header-test-1: "x-unique-id: 0A810258:BF7E_0A830013:0050_60F9E7A7_0001:0041"
        operator_test.go:1914: failed to read output from pod unique-id-header-test-2: container "curl" in pod "unique-id-header-test-2" is waiting to start: ContainerCreating
        operator_test.go:1914: failed to read output from pod unique-id-header-test-2: container "curl" in pod "unique-id-header-test-2" is waiting to start: ContainerCreating
        operator_test.go:1926: found x-unique-id header from pod unique-id-header-test-2: "x-unique-id: 0A800205:B608_0A830013:0050_60F9E7A9_0002:0041"
        operator_test.go:1914: failed to read output from pod unique-id-header-test-3: container "curl" in pod "unique-id-header-test-3" is waiting to start: ContainerCreating
        operator_test.go:1914: failed to read output from pod unique-id-header-test-3: container "curl" in pod "unique-id-header-test-3" is waiting to start: ContainerCreating
        operator_test.go:1914: failed to read output from pod unique-id-header-test-3: container "curl" in pod "unique-id-header-test-3" is waiting to start: ContainerCreating
        operator_test.go:1926: found x-unique-id header from pod unique-id-header-test-3: "x-unique-id: 0A800206:9E7C_0A830013:0050_60F9E7AD_0001:004C"
        operator_test.go:1904: failed to delete pod openshift-ingress/unique-id-header-test-1: pods "unique-id-header-test-1" not found
        panic.go:632: deleted ingresscontroller uniqueid
    --- FAIL: TestUniqueIdHeader (59.49s)

The test code should be changed to tolerate not-found errors when deleting pods.  I'll make that change in a follow-up PR (but I won't rush to get that fix in unless we see this flake again).  

/test e2e-aws-operator

Miciah commented:
> `TestUniqueIdHeader` failed; looks like one of its test pods got deleted (garbage collected?) after the test was done with it but before the test got around to cleaning it up:
> 
> ```
> === RUN   TestUniqueIdHeader
>     operator_test.go:1914: failed to read output from pod unique-id-header-[test-1](https://issues.redhat.com/browse/test-1): container "curl" in pod "unique-id-header-[test-1](https://issues.redhat.com/browse/test-1)" is waiting to start: ContainerCreating
>     operator_test.go:1914: failed to read output from pod unique-id-header-[test-1](https://issues.redhat.com/browse/test-1): container "curl" in pod "unique-id-header-[test-1](https://issues.redhat.com/browse/test-1)" is waiting to start: ContainerCreating
>     operator_test.go:1926: found x-unique-id header from pod unique-id-header-[test-1](https://issues.redhat.com/browse/test-1): "x-unique-id: 0A810258:BF7E_0A830013:0050_60F9E7A7_0001:0041"
>     operator_test.go:1914: failed to read output from pod unique-id-header-[test-2](https://issues.redhat.com/browse/test-2): container "curl" in pod "unique-id-header-[test-2](https://issues.redhat.com/browse/test-2)" is waiting to start: ContainerCreating
>     operator_test.go:1914: failed to read output from pod unique-id-header-[test-2](https://issues.redhat.com/browse/test-2): container "curl" in pod "unique-id-header-[test-2](https://issues.redhat.com/browse/test-2)" is waiting to start: ContainerCreating
>     operator_test.go:1926: found x-unique-id header from pod unique-id-header-[test-2](https://issues.redhat.com/browse/test-2): "x-unique-id: 0A800205:B608_0A830013:0050_60F9E7A9_0002:0041"
>     operator_test.go:1914: failed to read output from pod unique-id-header-[test-3](https://issues.redhat.com/browse/test-3): container "curl" in pod "unique-id-header-[test-3](https://issues.redhat.com/browse/test-3)" is waiting to start: ContainerCreating
>     operator_test.go:1914: failed to read output from pod unique-id-header-[test-3](https://issues.redhat.com/browse/test-3): container "curl" in pod "unique-id-header-[test-3](https://issues.redhat.com/browse/test-3)" is waiting to start: ContainerCreating
>     operator_test.go:1914: failed to read output from pod unique-id-header-[test-3](https://issues.redhat.com/browse/test-3): container "curl" in pod "unique-id-header-[test-3](https://issues.redhat.com/browse/test-3)" is waiting to start: ContainerCreating
>     operator_test.go:1926: found x-unique-id header from pod unique-id-header-[test-3](https://issues.redhat.com/browse/test-3): "x-unique-id: 0A800206:9E7C_0A830013:0050_60F9E7AD_0001:004C"
>     operator_test.go:1904: failed to delete pod openshift-ingress/unique-id-header-[test-1](https://issues.redhat.com/browse/test-1): pods "unique-id-header-[test-1](https://issues.redhat.com/browse/test-1)" not found
>     panic.go:632: deleted ingresscontroller uniqueid
> --- FAIL: TestUniqueIdHeader (59.49s)
> ```
> 
> The test code should be changed to tolerate not-found errors when deleting pods. I'll make that change in a follow-up PR (but I won't rush to get that fix in unless we see this flake again).
> 
> /test e2e-aws-operator

Fixed in https://github.com/openshift/cluster-ingress-operator/pull/638/commits/8041f493255a466cc6e4b7ab30f9ebea1d9664c6.

Miciah commented:
Test:

    foo-[test-1]bar

That was a test.

Miciah commented:
Test:

    foo-[test-1]: bar

and

```
foo-[test-1]: bar
```

That was a test.

Miciah commented:
More tests:

    operator_test.go:1914: failed to read output from pod unique-id-header-[test-1](https://issues.redhat.com/browse/test-1): container "curl" in pod "unique-id-header-[test-1](https://issues.redhat.com/browse/test-1)" is waiting to start: ContainerCreating

```
operator_test.go:1914: failed to read output from pod unique-id-header-[test-1](https://issues.redhat.com/browse/test-1): container "curl" in pod "unique-id-header-[test-1](https://issues.redhat.com/browse/test-1)" is waiting to start: ContainerCreating
```

Miciah commented:
I meant to do this test:

    operator_test.go:1914: failed to read output from pod unique-id-header-test-1: container "curl" in pod "unique-id-header-test-1" is waiting to start: ContainerCreating

```
operator_test.go:1914: failed to read output from pod unique-id-header-test-1: container "curl" in pod "unique-id-header-test-1" is waiting to start: ContainerCreating
```
